Fix #3045: introduce draft of the end to end test for classifyHandler

pranavsid98 opened the pull request:
@anmolshkl @seanlip . Am I on the right track? I know that I'd need to be using asssertEquals, I'll be updating that in the next commit if I've started work on the right direction. Thanks.

anmolshkl commented:
@pranavsid98 no, you are not supposed to use the lda_string_classifier_performance_test file. The reason why I suggested that you look at the file is to see how the sample exploration with the training data is loaded and how everything works. I can suggest two test cases to you - (1) when the predicted answer group is correct i.e. you already know to which answer group the answer belongs and the predicted answer group is same, (2) when the predicted answer group is incorrect because the answer is totally different (or irrelevant) from the training data.

PS: Look at the other typical test cases to see how test cases are structured and what they look like.

pranavsid98 commented:
@anmolshkl. Oh. That's fine, I'll follow a similar approach to it then. Thank you.

pranavsid98 commented:
@anmolshkl. How does it look now? I still sort of don't get what the handler's for. reader_test.py has not even imported reader.py file. So, I don't see how Handlers are being used. Could you help out here?
Thanks

anmolshkl commented:
@pranavsid98 handlers are not tested in that way I think. Handlers handle the incoming request for some particular URL. So, to test a handler you'd need to "post" the data to that URL. Check out how other handlers are tested in reader_test.py in similar a manner.

pranavsid98 commented:
@anmolshkl . I think this is how Handler testing works. Its working perfectly and I checked it as well. Could you review?

anmolshkl commented on core/controllers/reader_test.py:
```diff
@@ -103,6 +106,59 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+    """Test the hander for classification"""
+
+    def setUp(self):
+        super(ClassifyHandlerTest,self).setUp()
```
You need a blank space before ",".

pranavsid98 replied:
Done.

anmolshkl commented on core/controllers/reader_test.py:
```diff
@@ -103,6 +106,59 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
             expect_errors=True)
         self.assertEqual(response.status_int, 200)
 
```
Leave two new lines before a class.

pranavsid98 replied:
Done.

anmolshkl commented on core/controllers/reader_test.py:
```diff
@@ -103,6 +106,59 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+    def setUp(self):
+        super(ClassifyHandlerTest,self).setUp()
+
+    def test_classification_handler(self):
```
Umm I think you need to make to cases - (1) to test that a correct answer is being classified correctly and the output answer group is correct, (2) to test that the output is default answer group given an incorrect answer. So, you can create two functions like - test_classification_handler_with_correct_answer and test_classification_handler_with_incorrect_answer . You can argue that this should be a part of test cases written for classifier services but I don't see that it would create any major difference.

anmolshkl replied:
@pranavsid98 open/unresolved comments like this one.

anmolshkl commented on core/controllers/reader_test.py:
```diff
@@ -103,6 +106,59 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+        yaml_path = os.path.join(os.path.dirname(os.path.realpath(__file__)),
+                                 '../tests/data/string_classifier_test.yaml')
+        doc_to_label = {}
+        with open(yaml_path, 'r') as yaml_file:
```
Perhaps we can move this to setup() (if we are really creating two different tests so that all the necessary setup could be done in one go).

pranavsid98 replied:
Makes sense. Done.

anmolshkl commented on core/controllers/reader_test.py:
```diff
@@ -103,6 +106,59 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+        answer='Permutations'
+        params = { 'old_state' : old_state, 'answer' : answer }
+        feconf.ENABLE_STRING_CLASSIFIER = True
+        res = self.post_json('/explorehandler/classify/%s' % exp_id,params)
```
This looks good but we also need to check the response too. Specifically, we need to see if the result (i.e. answer group) is correct or not. Also, it might have struck you that this test case doesn't have any assert statements, so what is it really testing? :)

pranavsid98 replied:
My bad. This test shows true because the handler is being run. I'll add asserts for the output as well.

anmolshkl commented on core/controllers/reader_test.py:
```diff
@@ -103,6 +106,59 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+                                doc_to_label[doc] = []
+                            doc_to_label[doc].append(label)
+        examples = [[doc, doc_to_label[doc]] for doc in doc_to_label]
+        docs_to_classify = [doc[0] for doc in examples]
```
Better to remove this variable if you are not using it.

pranavsid98 replied:
Done

anmolshkl commented on core/controllers/reader_test.py:
```diff
@@ -103,6 +106,59 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+            classifier_registry.ClassifierRegistry.get_classifier_by_id(
+                feconf.INTERACTION_CLASSIFIER_MAPPING['TextInput']))
+        classifier.train(examples)
+        classifier_dict = classifier.to_dict()
```
Unused variable.

pranavsid98 replied:
To be exact, the entire part isn't required I believe. This makes the import not required as well.

pranavsid98 commented on core/controllers/reader_test.py:
```diff
@@ -18,6 +18,9 @@
 from core.domain import exp_services
 from core.domain import param_domain
 from core.domain import rights_manager
+from core.domain import classifier_registry
```
Done.

pranavsid98 commented on core/controllers/reader_test.py:
```diff
@@ -103,6 +106,59 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+        #Testing the handler
+        self.login(self.VIEWER_EMAIL)
+        old_state = exploration.states['Home'].to_dict()
+        answer='Permutations'
```
Done

pranavsid98 commented on core/controllers/reader_test.py:
```diff
@@ -103,6 +106,59 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+        self.login(self.VIEWER_EMAIL)
+        old_state = exploration.states['Home'].to_dict()
+        answer='Permutations'
+        params = { 'old_state' : old_state, 'answer' : answer }
```
Done

pranavsid98 commented on core/controllers/reader_test.py:
```diff
@@ -18,6 +18,9 @@
+from core.domain import classifier_registry
+import yaml
+import os
 from core.tests import test_utils
```
Done

pranavsid98 commented:
I updated the test cases, removed unused variables and redundant code. Addressed all changes as well. Have a look at the second test case as I'm sure it needs to be classified to the default outcome, but it's being classified as 'Detected Unsure'.

anmolshkl commented on core/controllers/reader.py:
```diff
@@ -333,8 +333,8 @@ def post(self, unused_exploration_id):
         # The learner's parameter values.
-        params = self.payload.get('params')
-        params['answer'] = answer
+        #params = self.payload.get('params')
```
Perhaps we can remove both the lines.

pranavsid98 replied:
Done.

anmolshkl commented:
@pranavsid98 I checked out your code and turns out that for any unknown answer, the classifier is outputting "Detected unsure" instead of  "Did not detect. Try again?", which is the default outcome. The current implementation tries to normalize the probability over different labels and my guess is that it is messing up somewhere there. I tried digging up and found that the numbers generated by LDA classifier are quite suspicious and in many cases blatantly wrong! Even if you input "1", it shows a prediction confidence of 0.73 for the response "Detected factorial" (probably because one training input is "3 x 2 x 1"). 

@seanlip since LDA was implemented before I started working on this project, I had assumed that it was working properly. Thanks to the test that we caught the problem before rolling this feature out! Now, I'll try tuning the algorithm and I'll simultaneously try to implement a parallel algorithm for comparison purpose. If LDA doesn't seem like a good fit, we can always swap in another algorithm.

pranavsid98 commented:
@anmolshkl. That was exactly the query. This is an important issue right? I'll try to dig the problem up with the LDA as well. I'm gonna possibly try a different data set of sorts to see if this issue migrates. 
Is the end to end test for ClassifyHandler resolved though?

anmolshkl commented:
@pranavsid98 PTAL on all the open comments.

pranavsid98 commented:
I've addressed the issue. Thanks for LGTM.

seanlip commented on core/controllers/reader.py:
```diff
@@ -332,9 +332,6 @@ def post(self, unused_exploration_id):
         old_state = exp_domain.State.from_dict(self.payload.get('old_state'))
         # The learner's raw answer.
         answer = self.payload.get('answer')
-        # The learner's parameter values.
```
Why are we modifying this? I thought the point of this PR was just to add a test.

pranavsid98 replied:
Those parameter values weren't required. Leaving it like that caused Lint errors (unused variable). 
Other than this, the line params['answer'] = answer wasn't required either as the 'params' dict already had that key.

seanlip replied:
OK. (Surprised that it didn't get picked up before, but you're right -- thanks for explaining.)

seanlip commented on core/controllers/reader_test.py:
```diff
@@ -104,6 +107,55 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+        """Before the test, create an exploration_dict."""
+        super(ClassifyHandlerTest, self).setUp()
+
+        #Reading YAML exploration into a dictionary
```
Add space after "#" and put full stop at end. Ditto elsewhere.

pranavsid98 replied:
Done.

seanlip commented on core/controllers/reader_test.py:
```diff
@@ -104,6 +107,55 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+
+        self.login(self.VIEWER_EMAIL)
+
+
```
Drop extra newline.

pranavsid98 replied:
Done

seanlip commented on core/controllers/reader_test.py:
```diff
@@ -104,6 +107,55 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+        self.signup(self.VIEWER_EMAIL, self.VIEWER_USERNAME)
+
+        # Load demo exploration.
+        exp_id = '0'
```
If you define exp_id you might as well use it below, rather than continue hardcoding '0'?

pranavsid98 replied:
Done.

seanlip commented on core/controllers/reader_test.py:
```diff
@@ -104,6 +107,55 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+        old_state = exploration.states['Home'].to_dict()
+        answer = 'Permutations'
+        params = {'old_state' : old_state, 'answer' : answer}
+        feconf.ENABLE_STRING_CLASSIFIER = True
```
Don't do this; there is a construct along the lines of `with self.swap(feconf, '...')` that you should use instead. See other tests for details.

pranavsid98 replied:
Done

seanlip commented on core/controllers/reader_test.py:
```diff
@@ -104,6 +107,55 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+        exp_services.load_demo('0')
+
+        #Creating the exploration domain object.
+        self.yaml_dict['id'] = 20
```
why?

pranavsid98 replied:
There was a key error while using from_dict method. It needed an ID key which wasn't present in the dict. The handler specified that it needed an ID for just @require_playable function defined before the handler. So, I used a random ID. Is there something else I should do there?

seanlip replied:
Don't we have an exploration.from_yaml() method?

pranavsid98 replied:
Ooooh, I think we do. That reduces just so much work. I'll get to that now.

seanlip commented on core/controllers/reader_test.py:
```diff
@@ -104,6 +107,55 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+        exploration = exp_domain.Exploration.from_dict(self.yaml_dict)
+
+        #Testing the handler for a correct answer.
+        old_state = exploration.states['Home'].to_dict()
```
old_state_dict (distinguish from the domain object, since it's a dict)

pranavsid98 replied:
I used it as old_state because that's the key that was used in the ClassifyHandler class. So, if I change it here, I'd need to change it in the reader.py file as well. Should I?

seanlip replied:
There is a difference between the object you have here and the object in ClassifyHandler. Try logging their types.

pranavsid98 replied:
Done

seanlip commented on core/controllers/reader_test.py:
```diff
@@ -104,6 +107,55 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+        #Testing the handler for a correct answer.
+        old_state = exploration.states['Home'].to_dict()
+        answer = 'Permutations'
+        params = {'old_state' : old_state, 'answer' : answer}
```
Load an exploration in a dev server, play it, see what params actually contains, and use that.

pranavsid98 replied:
I dont quite get how I'd do this. Could you please help me here?

seanlip replied:
If the Python backend handlers are being called when you're playing with it, then it should be possible for you to insert logging statements in those handlers, and see the logs in the server output in the terminal.

pranavsid98 replied:
Great. Output of self.payload is : 
{u'params': {u'answer': 100}, u'old_state_name': u'Estimate 100', u'answer': 100, u'rule_spec_index': 0, u'version': 1, u'answer_group_index': 0}
Now, we can modify this accordingly.

pranavsid98 replied:
Done

seanlip commented:
> @seanlip since LDA was implemented before I started working on this project, I had assumed that it was working properly. Thanks to the test we caught the problem before rolling this feature out! Now, I'll try tuning the algorithm and I'll simultaneously try to implement a parallel algorithm for comparison purpose. If LDA doesn't seem like a good fit, we can always swap in another algorithm.

Thanks for noticing that. I actually thought it was working properly, but please do investigate, since I haven't verified it closely. It would be a good idea to augment the test suite for that classifier to catch cases that have issues like this!

seanlip commented on core/controllers/reader_test.py:
```diff
@@ -104,6 +107,58 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
+            self.logout()
+
+
+
```
There should only be two newlines above a top-level classdef.

pranavsid98 replied:
Done

pranavsid98 commented:
I think everything's good now. I also want to let you know that the ClassifyHandler wasn't being called in the demo explorations. I found value of the 'params' by logging it in AnswerSubmittedEventHandler. When exactly does ClassifyHandler gets called?

seanlip commented on core/controllers/reader.py:
```diff
@@ -329,7 +329,8 @@ def post(self, unused_exploration_id):
         # A domain object representing the old state.
-        old_state = exp_domain.State.from_dict(self.payload.get('old_state'))
+        old_state = exp_domain.State.from_dict(self.payload.get(
+            'old_state_dict'))
```
Why are you changing the name of the field that's coming in via the payload? I expect this to break stuff...

pranavsid98 replied:
Oh. My bad. Done. Changed it back.

pranavsid98 commented:
@seanlip. I've got the training interface running properly, I've made the required changes. And I've loaded the classifier_demo_exploration. None of the FuzzyMatches rule types are getting matched. It checks for only 'contains' happy and everything else seems to go to default outcome. And nop, ENABLE_STRING_CLASSIFIER isn't helping.

seanlip commented:
Ah, thanks @pranavsid98. I can't really take a look this week, but will pass this to @anmolshkl for further investigation. In the meantime, purely from a code perspective, the backend tests look good to me, so I'll give LGTM and defer to @anmolshkl for final approval.

anmolshkl commented:
@pranavsid98 if you have already enabled the training interface by setting SHOW_TRAINABLE_UNRESOLVED_ANSWERS to True, I don't understand why it is not being called. Can you try to place debug points (or something similar, whatever works for you) and see what's really happening there? As far as I remember, calls were being made from the front-end to the ClassifyHandler last time I checked it out. 

@seanlip since the issue (of LDA not working as we had expected) is making the test fail, I don't think we can merge this PR. I think we might have to see what's going wrong here. @pranavsid98 do you want to look into this?

pranavsid98 commented:
@anmolshkl. Yes, I'd like to look into this. But then, the problem doesn't have anything to do with this PR does it? This test can be changed to check the default outcome as soon as we get the LDA working. This test is working fine as of now as it is not testing the default outcome. 
Oh, and I didn't set 'SHOW_TRAINABLE_UNRESOLVED_ANSWERS'. Didn't know we needed that. I'll try with that and update.
Would you be opening an issue for the LDA bug?

anmolshkl commented:
@pranavsid98 you need to set SHOW_TRAINABLE_UNRESOLVED_ANSWERS to True in order to enable the training interface. The current editor (in production) doesn't support training interface and it can be turned on in dev mode only by setting the variable to True.

Also, we cannot merge a partially failing test case into develop. But I think the point of this test should be to only check whether the handler is working or not and not whether classifier is performing properly or not (for which we should have a separate performance test case, ideally). In that case, we can do away with the scenario where the answer should have a default outcome. @seanlip what's your opinion?

seanlip commented:
@anmolshkl it's your call, I don't have an opinion. You have a better idea of the overall test suite than I do; I think the main thing is to ensure that all the different parts get covered before the functionality goes live.

pranavsid98 commented:
@anmolshkl . If that's the case, then should I just remove that test case and make a commit?

anmolshkl commented:
@seanlip thanks!

@pranavsid98 yep, its better to remove the second test case. Other than that, this PR LGTM!
 FYI, it's better to keep different test scenarios in different methods. This helps you when you are executing the test suite as it's easier to keep track of the different scenarios that are failing or passing.

pranavsid98 commented:
@anmolshkl. Can we merge this PR now?

anmolshkl commented:
@pranavsid98 yep, I was waiting for a flaky test to pass in the Travis build.
